File: CSharpCodeBase/aiactions/snowtroll/ai_target_search.lua
local Class = require 'utils.hump.class'
local AINode = require 'utils.ailib.ainode'

local AI_TargetSearch = Class({__includes=AINode, init = function (self, entity, callback)
  AINode.init(self, entity)
  self.callback = callback
  self:ResetTarget()
end})

function AI_TargetSearch:ResetTarget()
  self.currentTarget=nil
end

function AI_TargetSearch:Visit()
  if self.status == NODE_READY then
    self:ResetTarget()
    local possibleTargets = self.entity.vision:GetVisibleEntitiesByRSTag("alertness") or self.entity.vision:GetVisibleEntitiesByRSTag("enemy")
    self.currentTarget = possibleTargets[1]
    if self.currentTarget ~= nil then
      self.status = NODE_RUNNING
    else
      self.status = NODE_FAILURE
    end
  end
  
  if self.status == NODE_RUNNING then
    self.entity.goalProsecution = self.currentTarget
    self.entity.lead = self.entity
    self.entity._screamTrigger = true
    self.entity._screamTarget = self.currentTarget
    self.status = NODE_SUCCESS
    self.callback(self.entity, self)
  end
end

return AI_TargetSearch

File: CSharpCodeBase/aiactions/snowtroll/ai_target_search.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class ai_target_search :ainode{
 var AI_TargetSearch = Class({__includes=AINode, init = public void (self, entity, callback){
   AINode.init(self, entity);
   this.callback = callback;
   self:ResetTarget();
 }})
 public void AI_TargetSearch:ResetTarget(){
   this.currentTarget=nil;
 }
 public void AI_TargetSearch:Visit(){
   if(this.status == NODE_READY  ){ 
     self:ResetTarget();
     var possibleTargets = this.entity.vision:GetVisibleEntitiesByRSTag("alertness")  ||  this.entity.vision:GetVisibleEntitiesByRSTag("enemy");
     this.currentTarget = possibleTargets[1];
     if(this.currentTarget != null  ){ 
       this.status = NODE_RUNNING;
     }else{
       this.status = NODE_FAILURE;
     }
   }
   
   if(this.status == NODE_RUNNING  ){ 
     this.entity.goalProsecution = this.currentTarget;
     this.entity.lead = this.entity;
     this.entity._screamTrigger = true;
     this.entity._screamTarget = this.currentTarget;
     this.status = NODE_SUCCESS;
     this.callback(this.entity, self);
   }
 }
 return AI_TargetSearch}}